Refactor GraphCache to support multiple types for the same topic.

Yadunund opened the pull request:
This PR
- Updates the `GraphCache` implementation to support multiple types over the same topic name.
- Previously we were not keeping track of multiple pub/subs in the same node that publish over the same topic and let alone with different topic types. This PR addresses both these problems as well.
- It also reduces code duplication when adding/removing pub/subs by relying on lambdas that are reused for both pub/subs.
- Adds data structures and methods (see `liveliness_utils.hpp`) to manage token generation and conversion to C++ types. No longer have any magic string literal comparisons or custom key generation for unordered_maps.

Yadunund committed:
use std::find_if to find matching pubs

Signed-off-by: Yadunund <yadunund@openrobotics.org>

Yadunund committed:
Refactor to support multiple types

Signed-off-by: Yadunund <yadunund@openrobotics.org>

gavanderhoorn commented:
> This PR updates the GraphCache implementation to support multiple types over the same topic name.

I'm just watching this repository, so likely out-of-the-loop a bit, but what would be a use-case for this? Doesn't that break the "strong typing" of ROS topics?

clalancette commented:
> I'm just watching this repository, so likely out-of-the-loop a bit, but what would be a use-case for this? Wouldn't that break the "strong typing" of ROS topics?

This is a supported use-case today:

Terminal 1:
```
$ ros2 topic pub -r 1 /foo std_msgs/msg/String '{data: hello}'
```

Terminal 2:
```
$ ros2 topic pub -r 1 /foo std_msgs/msg/Int32 '{data: 3}'
```

Terminal 3:
```
$ ros2 topic list -t
/foo [std_msgs/msg/Int32, std_msgs/msg/String]
/parameter_events [rcl_interfaces/msg/ParameterEvent]
/rosout [rcl_interfaces/msg/Log]
```

Is it a good idea?  No.  But since the aim is to keep most of the semantics the same with `rmw_zenoh_cpp`, we should support it.

gavanderhoorn commented:
> This is a supported use-case today

yes, I know ..

> Is it a good idea? No.

Indeed.

Yadunund committed:
Refactor liveliness token generation

Signed-off-by: Yadunund <yadunund@gmail.com>

Yadunund committed:
Map topic types to TopicData

Signed-off-by: Yadunund <yadunund@gmail.com>

Yadunund committed:
Store map of types in graph_topics_

Signed-off-by: Yadunund <yadunund@gmail.com>

Yadunund commented:
> > I'm just watching this repository, so likely out-of-the-loop a bit, but what would be a use-case for this? Wouldn't that break the "strong typing" of ROS topics?
> 
> This is a supported use-case today:
> 
> Terminal 1:
> 
> ```
> $ ros2 topic pub -r 1 /foo std_msgs/msg/String '{data: hello}'
> ```
> 
> Terminal 2:
> 
> ```
> $ ros2 topic pub -r 1 /foo std_msgs/msg/Int32 '{data: 3}'
> ```
> 
> Terminal 3:
> 
> ```
> $ ros2 topic list -t
> /foo [std_msgs/msg/Int32, std_msgs/msg/String]
> /parameter_events [rcl_interfaces/msg/ParameterEvent]
> /rosout [rcl_interfaces/msg/Log]
> ```
> 
> Is it a good idea? No. But since the aim is to keep most of the semantics the same with `rmw_zenoh_cpp`, we should support it.

@clalancette this should work with PR now. Could you give it a spin?

Yadunund committed:
fix bug in deleting and counting topics

Signed-off-by: Yadunund <yadunund@gmail.com>

Yadunund committed:
Cleanup liveliness_utils

Signed-off-by: Yadunund <yadunund@gmail.com>

clalancette committed:
Style cleanups.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette committed:
Cleanup some of the code in the liveliness.

Also fix a couple of bugs where we were mixing up subscriptions
and publications, leading to incorrect counts.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette committed:
Remove Entity::node_info.

Because it returned a reference, it is difficult to
tell the lifetime of it.  It would also be hard to add
in locking in the future.  Since we are reaching through
it in most cases, just return some data directly.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette commented on rmw_zenoh_cpp/src/detail/liveliness_utils.cpp:
```diff
@@ -0,0 +1,362 @@
+const std::optional<TopicInfo> & Entity::topic_info() const
+{
+  return this->topic_info_;
+}
```
I'm not a huge fan of APIs like this, that return a reference to something inside of the class.

There are two (related) reasons:
1.  It becomes difficult to tell what the lifetime of this returned value is.  If the caller took a reference, but then this class gets destroyed, it is now holding onto a dangling reference.
2.  If we want to add in locking around `topic_info_`, for instance, it becomes tricky; we have to expose the locking mechanism outside of the class to all callers, and then all callers have to remember to actually lock it.

You can see that I actually changed `node_info` above to three separate APIs that just return the needed data without references.  That is much safer, in my opinion, and I think we should do a similar thing for `topic_info()`, `type()`, and `keyexpr()` in this PR.

Yadunund replied:
Thanks for the feedback! I've pushed https://github.com/ros2/rmw_zenoh/pull/70/commits/4488b976fa205a1ec2c8f98b538b49d96934d185. I hope it's fine to let `topic_info()`still return an `std::optional<TopicInfo>`. Else if we split into 3 APIs, we will need to check all 3 params are set each time we want to parse necessary data for pub/subs. To prevent large copies of strings each time, I've updated the implementation to make one copy of `topic_info()` and access the required fields where ever possible.  

> Besides that, I'm noticing segmentation faults and errors when I shut nodes down with this. I suspect that we are trying to process some liveliness tokens while we are shutting down the context, but I'm not sure. I think we should probably address that before merging this in.

Do you have any commands to reproduce this behavior? I wasn't able to do so. But I did encounter this problem early on when I had the `ros2 daemon` running from an older commit of this branch. After `ros2 daemon stop` and rebuilding I never saw that again.

clalancette committed:
Remove some uses of auto.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette reviewed (changes requested):
Overall, this is looking pretty good.

Inline, I've left one major piece of feedback about APIs.

Besides that, I'm noticing segmentation faults and errors when I shut nodes down with this.  I suspect that we are trying to process some liveliness tokens while we are shutting down the context, but I'm not sure.  I think we should probably address that before merging this in.

Yadunund committed:
Return copies from Entity

Signed-off-by: Yadunund <yadunund@gmail.com>

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -32,491 +33,347 @@
-  token += topic + "/" + type + "/" + qos;
-  return token;
-}
+  std::optional<liveliness::Entity> valid_entity = liveliness::Entity::make(keyexpr);
+  if (!valid_entity.has_value()) {
+    // Error message has already been logged.
+    return;
+  }
```
Regarding error handling.

When the entity can't be created, the error msg is set and then it is returned. However, when the method (`parse_put`) is being called there is no way to know it failed from the caller's perspective.

Yadunund replied:
It can be argued how to deal with errors with discovery and whether that should throw and exception or continue the process if the data transport is still working. Maybe something we can iterate on as things get clearer?

Yadunund replied:
Opened a ticket to work on in the next iteration https://github.com/ros2/rmw_zenoh/issues/72

francocipollone commented on rmw_zenoh_cpp/src/detail/liveliness_utils.cpp:
```diff
@@ -0,0 +1,362 @@
+        "rmw_zenoh_cpp",
+        "Received liveliness token for non-node entity without required parameters.");
+      return std::nullopt;
+    }
```
nit: Just for consistency, should we check here if we have strictly more than 8 parts? We shouldn't reach that part though.

Yadunund replied:
I'd be in favor of not having any magic numbers like `8` anywhere in the code. Maybe we can update the `Entity` API to return the number of parts for a given `EntityType`.

Yadunund replied:
Ticketed to work on in the next iteration https://github.com/ros2/rmw_zenoh/issues/71

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -32,491 +33,347 @@
+  if (ns_it == graph_.end()) {
+    NodeMap node_map = {
+      {entity.node_name(), make_graph_node(entity)}};
+    graph_.insert(std::make_pair(entity.node_namespace(), std::move(node_map)));
```
nit: using `emplace` instead of `insert` might be useful to not rely on the compiler for optimization in the construction/copy of the object

Emplace creates the object in-place instead of moving it with insert if possible.

https://en.cppreference.com/w/cpp/container/unordered_map/emplace

Yadunund replied:
Done https://github.com/ros2/rmw_zenoh/pull/70/commits/357f6588fab1f2f9192e7be418d9f1c3b4956da5

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -32,491 +33,347 @@
-    RCUTILS_SET_ERROR_MSG("The zenoh session is invalid.");
-    return false;
-  }
+      const liveliness::TopicInfo topic_info = entity.topic_info().value();
```
```suggestion
      const liveliness::TopicInfo & topic_info = entity.topic_info().value();
```

Yadunund replied:
We want to create a copy so that the `graph_topic_data` object created in the next line will not have a dangling reference to `TopicInfo` after the lambda goes out of scope.

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -32,491 +33,347 @@
+        );
+        graph_topics_[topic_info.name_] = GraphNode::TopicDataMap{
+          {topic_info.type_, topic_data_ptr}
+        };
```
Isn't it the same as `topic_data_map`?

```suggestion
        graph_topics_[topic_info.name_] = topic_data_map;
```

Yadunund replied:
Done https://github.com/ros2/rmw_zenoh/pull/70/commits/357f6588fab1f2f9192e7be418d9f1c3b4956da5

francocipollone commented on rmw_zenoh_cpp/src/detail/graph_cache.hpp:
```diff
@@ -137,20 +122,14 @@ class GraphCache final
+  NamespaceMap graph_ = {};
+
+  // Optimize topic lookups across the graph.
+  GraphNode::TopicMap graph_topics_ = {};
```
Just for documentation's sake. Without this optimized collection for topic lookups, does the time for the queries(for topics) go up considerably? Just wondering why solving the query on demand isn't a chance. Taking into account that keeping the double source of truth might be error-prone.

Yadunund replied:
With the cache, lookups to check if a certain topic exists in the graph have `O(1)` complexity on average. Else if we have to iterate through `graph_cache_`, it will be `O(n*n'*n'')` where we have to iterate through `n` namepaces, `n'` nodes within each namespace and potentially `n''` topics per node.

francocipollone reviewed (commented):
Great stuff!  :rocket: 

I left some minor comments 

I tested locally running demo nodes:

```
(docker) franco@:~/ros_dev_ws$ ros2 topic list -v
Published topics:
 * /chatter [std_msgs/msg/String] 1 publisher
 * /parameter_events [rcl_interfaces/msg/ParameterEvent] 3 publishers
 * /rosout [rcl_interfaces/msg/Log] 3 publishers

Subscribed topics:
 * /chatter [std_msgs/msg/String] 1 subscriber
 * /parameter_events [rcl_interfaces/msg/ParameterEvent] 2 subscribers
```

I haven't experienced any error when shutting down the nodes

Yadunund committed:
Address feedback

Signed-off-by: Yadunund <yadunund@gmail.com>

Yadunund commented:
Thanks for the review @clalancette and @francocipollone! I've addressed the blocking feedback and opened tickets for the others. Will merge this in now to continue impl of introspection methods.
